src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/FileBasedProgramsProjectSystem.cs

```diff
@@ -0,0 +1,154 @@
+
+    public async ValueTask<TextDocument?> AddMiscellaneousDocumentAsync(DocumentUri uri, SourceText documentText, string languageId, ILspLogger logger)
+    {
+        var documentFilePath = GetDocumentFilePath(uri);
+
+        // https://github.com/dotnet/roslyn/issues/78421: MetadataAsSource should be its own workspace
+        if (_metadataAsSourceFileService.TryAddDocumentToWorkspace(documentFilePath, documentText.Container, out var documentId))
+        {
+            var metadataWorkspace = _metadataAsSourceFileService.TryGetWorkspace();
+            Contract.ThrowIfNull(metadataWorkspace);
+            var document = metadataWorkspace.CurrentSolution.GetRequiredDocument(documentId);
+            return document;
+        }
+
+        var primordialDoc = AddPrimordialDocument(uri, documentText, languageId);
+        Contract.ThrowIfNull(primordialDoc.FilePath);
+
+        if (uri.ParsedUri?.IsFile is true
+            && primordialDoc.Project.Language == LanguageNames.CSharp
+            && GlobalOptionService.GetOption(LanguageServerProjectSystemOptionsStorage.EnableFileBasedPrograms))
+        {
+            await BeginLoadingProjectWithPrimordialAsync(primordialDoc.FilePath, primordialProjectId: primordialDoc.Project.Id);
+        }
+        else
+        {
+            await TrackPrimordialOnlyProjectAsync(primordialDoc.FilePath, primordialDoc.Project.Id);
+        }
+
+        return primordialDoc;
+
+        TextDocument AddPrimordialDocument(DocumentUri uri, SourceText documentText, string languageId)
+        {
+            var languageInfoProvider = _lspServices.GetRequiredService<ILanguageInfoProvider>();
+            if (!languageInfoProvider.TryGetLanguageInformation(uri, languageId, out var languageInformation))
+            {
+                Contract.Fail($"Could not find language information for {uri} with absolute path {documentFilePath}");
+            }
+
+            var workspace = Workspace;
+            var sourceTextLoader = new SourceTextLoader(documentText, documentFilePath);
+            var projectInfo = MiscellaneousFileUtilities.CreateMiscellaneousProjectInfoForDocument(
+                workspace, documentFilePath, sourceTextLoader, languageInformation, documentText.ChecksumAlgorithm, workspace.Services.SolutionServices, []);
+
+            // Ensure that syntax errors for `#:` directives are not reported on the primordial project.
+            Contract.ThrowIfNull(projectInfo.ParseOptions);
+            projectInfo = projectInfo.WithParseOptions(projectInfo.ParseOptions.WithFeatures([new("FileBasedProgram", "true")]));
+
+            ProjectFactory.ApplyChangeToWorkspace(workspace => workspace.OnProjectAdded(projectInfo));
+
+            // https://github.com/dotnet/roslyn/pull/78267
+            // Work around an issue where opening a Razor file in the misc workspace causes a crash.
```

Comment:
Not sure why this comment is here -- it's not a workaround...it's how the razor code works...?

Reply:
This was motivated by https://github.com/dotnet/roslyn/pull/78488#discussion_r2078655815.

Reply:
Potentially what we should do in the future is map language information to kinds of documents explicitly.  But that shouldn't be part of this PR.